### PR TITLE
Fix encoding a struct with a sub-record field default value

### LINKED ISSUE #321: Encoding a struct with a sub-record field default value failed
My understanding of Avro is that fields of type ```Record``` support default values only if the sub-record is defined separately (not as a nested record).

I tried the following scenario:

```go 
func TestDebug(t *testing.T) {
	_ = avro.MustParse(`{
		"name": "service.A",
		"type": "record",
		"fields": [{
			"name": "a",
			"type": "string"
		}]
	}`)
	schema := avro.MustParse(`{
		"name": "service.B",
		"type": "record",
		"fields": [{
			"name": "a",
			"type": "service.A",
			"default": {"a": "a value"}
		},{
			"name": "b",
			"type": "string"
		}]
	}`)

	type A struct {
		A string `avro:"a"`
	}

	type B struct {
		B string `avro:"b"`
	}

	b, _ := avro.Marshal(schema, B{B: "b value"})

	type B1 struct {
		B string `avro:"b"`
		A A      `avro:"a"`
	}

	b1 := B1{}
	_ = avro.Unmarshal(schema, b, &b1)

	if want, got := (A{A: "a value"}), b1.A; want != got {
		t.Errorf("%v,%v must be equals", want, got)
	}
}

```

<details>
<summary>It panics with the STACK TRACE (unfold)</summary>

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x1 pc=0x100ee7b]

goroutine 6 [running]:
testing.tRunner.func1.2({0x11c9020, 0x137f0e0})
        /usr/local/go/src/testing/testing.go:1545 +0x238
testing.tRunner.func1()
        /usr/local/go/src/testing/testing.go:1548 +0x397
panic({0x11c9020?, 0x137f0e0?})
        /usr/local/go/src/runtime/panic.go:914 +0x21f
reflect.mapaccess(0xc00004e460?, 0xc00010d930?, 0xc00010d9b8?)
        /usr/local/go/src/runtime/map.go:1350 +0x13
github.com/modern-go/reflect2.(*UnsafeMapType).UnsafeGetIndex(...)
        /Users/mac/go/pkg/mod/github.com/modern-go/reflect2@v1.0.2/unsafe_map.go:101
github.com/hamba/avro/v2.(*recordMapEncoder).Encode(0xc00006a880, 0xc00010be30, 0xc000029040)
        /Users/mac/dev/hamba_avro/codec_record.go:337 +0x107
github.com/hamba/avro/v2.(*structEncoder).Encode(0x13882b0?, 0xc000027dc0, 0xc000029040)
        /Users/mac/dev/hamba_avro/codec_record.go:204 +0xa7
github.com/hamba/avro/v2.(*Writer).WriteVal(0xc000029040, {0x124ce48, 0xc00011c370}, {0x11cff60, 0xc000027dc0})
        /Users/mac/dev/hamba_avro/codec.go:63 +0xed
github.com/hamba/avro/v2.(*frozenConfig).Marshal(0xc00012c080, {0x124ce48, 0xc00011c370}, {0x11cff60, 0xc000027dc0})
        /Users/mac/dev/hamba_avro/config.go:126 +0x5a
github.com/hamba/avro/v2.Marshal(...)
        /Users/mac/dev/hamba_avro/encoder.go:36
github.com/ln80/event-store/avro/v2.TestDebug(0xc0001269c0)
        /Users/mac/dev/larcade/event-store/avro/v2/draft_test.go:355 +0xd3
testing.tRunner(0xc0001269c0, 0x12146d0)
        /usr/local/go/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 1
        /usr/local/go/src/testing/testing.go:1648 +0x3ad
```
</details>

I dug deeper into the code, and it seems to me that ``` *reflect2.UnsafeMapType.UnsafeGetIndex ``` receives the wrong argument:

https://github.com/hamba/avro/blob/58cadd08715cb07eb2fa3e19b1a31bb0c0fbd96d/codec_record.go#L328-L332

``ptr`` should be an unsafe pointer of a map’s pointer: ```unsafe.Pointer(&map[string]any)```

### PR DESCRIPTION
**Pull request by redaLaanait**
Fixes #321

### CODE REVIEW

File: codec_record.go [hunk 1]
```diff
@@ -327,7 +333,8 @@ type recordMapEncoder struct {
 
 func (e *recordMapEncoder) Encode(ptr unsafe.Pointer, w *Writer) {
 	for _, field := range e.fields {
-		valPtr := e.mapType.UnsafeGetIndex(ptr, reflect2.PtrOf(field))
+		key := field.name
+		valPtr := e.mapType.UnsafeGetIndex(ptr, reflect2.PtrOf(&key))
```
Comment (nrwiersma) [inline codec_record.go:337]: This is not really doing anything. The reason is the way Go lays structs out in memory. When you have an unsafe pointer to a struct, you actually have a pointer to the first field in the struct, in this case `name`.

File: codec_record.go [hunk 2]
```diff
@@ -172,9 +172,15 @@ func encoderOfStruct(cfg *frozenConfig, schema Schema, typ reflect2.Type) ValEnc
 		}
 
 		defaultType := reflect2.TypeOf(def)
+
+		defaultEncoder := encoderOfType(cfg, field.Type(), defaultType)
+		if defaultType.LikePtr() {
+			defaultEncoder = &onePtrEncoder{defaultEncoder}
+		}
+
```
Comment (nrwiersma) [inline codec_record.go:180]: Nit: This is all the same block as it is a single sequence.

```suggestion
		defaultEncoder := encoderOfType(cfg, field.Type(), defaultType)
		if defaultType.LikePtr() {
			defaultEncoder = &onePtrEncoder{defaultEncoder}
		}
```

File: encoder_record_test.go [hunk 3]
```diff
@@ -139,6 +139,43 @@ func TestEncoder_RecordStructPartialWithNullDefault(t *testing.T) {
 	assert.Equal(t, []byte{0x00, 0x06, 0x66, 0x6f, 0x6f}, buf.Bytes())
 }
 
+func TestEncoder_RecordStructPartialWithSubRecordDefault(t *testing.T) {
+	defer ConfigTeardown()
+
+	// register the sub record as separate schema; nested records do not support default value
```
Comment (nrwiersma) [inline encoder_record_test.go:145]: This comment is not correct, and functionally in the library, there is no difference between this and it being a single schema.

File: encoder_record_test.go [hunk 4]
```diff
@@ -410,6 +447,44 @@ func TestEncoder_RecordMapWithDefault(t *testing.T) {
 	assert.Equal(t, []byte{0x36, 0x06, 0x66, 0x6f, 0x6f}, buf.Bytes())
 }
 
+func TestEncoder_RecordMapWithSubRecordDefault(t *testing.T) {
+	defer ConfigTeardown()
+
+	// register the sub record as separate schema; nested records do not support default value
```
Comment (nrwiersma) [inline encoder_record_test.go:453]: Same as above.

**Review by nrwiersma — CHANGES_REQUESTED** [review]
Nice catch. A couple comments.

File: codec_record.go (hunk 1 above)
Comment (redaLaanait) [reply codec_record.go:337]: You're right! It took me a while to figure out this behavior. I decided to add  ```key``` variable, regardless, to make the code clearer (IMO) and add some robustness; as you mentioned, the current implementation might break if ```mapEncoderField``` changes its first field.

File: encoder_record_test.go (hunk 3 above)
Comment (redaLaanait) [reply encoder_record_test.go:145]: You're right! I misread an old answer in stackOverflow...

File: codec_record.go (hunk 1 above)
Comment (redaLaanait) [reply codec_record.go:337]: I'll undo this change

**Review by nrwiersma — APPROVED** [review]
LGTM 🎉  Thanks for the contribution

**Pull request merged**